taxonomy.py

```diff
@@ -0,0 +1,493 @@
+        m = Metric(*args, **kwargs)
+        m.parent = self
+        self.metrics.append(m)
+        return m
+    
+    def check_solved(self):
+        if all(m.solved for m in self.metrics + self.subproblems):
+            self.solved = True
+            for p in self.superproblems:
+                p.check_solved()
+    
+    def __str__(self):
+        return "Problem({0})".format(self.name)
+    
+    def print_structure(self, indent=0):
+        print indent * " " + str(self)
+        for m in self.metrics:
+            print (indent + 4) * " " + str(m)
+        for p in self.subproblems:
+            p.print_structure(indent + 4)
+    
+    def tables(self):
+        return render_tables(self.metrics)
+
+mpl.rcParams["legend.fontsize"] = u"x-small"
+mpl.rcParams["xtick.labelsize"] = u"xx-small"
+mpl.rcParams["ytick.labelsize"] = u"x-small"
+
+class Metric:
+    def __init__(self, name, url=None, solved=False, notes="", scale=linear, target=None, target_source=None,
+                 parent=None, changeable=False, axis_label=None, target_label=None, data_url=None):
+        self.name = name
+        self.measures = []
+        self.solved = solved
+        self.url = url
+        self.notes = notes
+        self.scale = scale
+        self.target = target
+        self.target_source = target_source # Source for human-level performance number
+        self.changeable = changeable # True if a metric changes over time
+        self.graphed = False
+        global metrics
+        metrics[name] = self
+        self.parent = parent
+        self.target_label = target_label
+        self.axis_label = (     axis_label            if axis_label 
+                           else self.scale.axis_label if hasattr(self.scale, "axis_label") 
+                           else self.name)
+        # primarily used by the table() method
+        self.data_url = self.find_edit_url() if data_url is None else data_url
+        
+    def __str__(self):
+        solved = "SOLVED" if self.solved else "?" if not self.target else "not solved"
+        return "{0:<60}{1}".format("Metric(%s)" % self.name, solved)
+        
+    def measure(self, *args, **kwargs):
+        try:
+            m = Measurement(*args, **kwargs)
+        except AssertionError:
+            print "WARNING, failed to create measurement", args, kwargs
+            traceback.print_exc()
+            return None
+        m.metric = self
+        if self.target:
+            if self.target_source == m.url and self.target == m.value:
+                print "Skipping apparent human performance (target_source) paper", m.url
+                return None
+            if self.scale.improvement(self.target, m.value) >= 0:
+                self.solved = True
+                self.parent.check_solved()
+        self.measures.append(m)
+        return m
+
+    def find_edit_url(self):
+        "Some magic hackery to find what file and line number a Metric was defined on and produce an edit link"
```

Comment:
Explain that you intentionally trigger the exception to extract the line number.